### PR TITLE
exporter.koplugin: send the right mimetype when sharing text.

### PR DESCRIPTION
**Pull request by pazos**
Untested.

Pinging @uroybd: https://github.com/koreader/koreader/pull/9153#issuecomment-1143857139

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/9194)

### CODE REVIEW

**Comment by uroybd** [comment]
Will test it by Monday. :)

File: plugins/exporter.koplugin/target/json.lua [hunk 1]
```diff
@@ -1,10 +1,9 @@
-local Device = require("device")
 local rapidjson = require("rapidjson")
 
 -- json exporter
 local JsonExporter = require("base"):new {
     name = "json",
-    shareable = Device:canShareText(),
+    mimetype = "application/json",
```
[the remark this answers is not in the document]
Comment (pazos) [reply plugins/exporter.koplugin/target/json.lua:6]: Okay, no problem. 

The proper mimetype is [application/json](https://www.iana.org/assignments/media-types/application/json) but I guess google has the final word on its framework rules :p

File: plugins/exporter.koplugin/target/markdown.lua [hunk 2]
```diff
@@ -136,7 +136,7 @@ end
 
 function MarkdownExporter:share(t)
     local content = md.prepareBookContent(t, self.settings.formatting_options, self.settings.highlight_formatting) .. "\n\n_Generated at: " .. self:getTimeStamp() .. "_"
-    Device:doShareText(content)
+    self:shareText(content)
```
[the remark this answers is not in the document]
Comment (pazos) [reply plugins/exporter.koplugin/target/markdown.lua:139]: Yes, that's the intended behaviour. I'm willing to merge any tested suggestion :)

File: plugins/exporter.koplugin/target/json.lua (hunk 1 above)
[the remark this answers is not in the document]
Comment (uroybd) [reply plugins/exporter.koplugin/target/json.lua:6]: Yeah, it is very weird. However, the rationale maybe (and not a good one) to allow `text/*` mimetype for JSON too.

[the remark this answers is not in the document]
Comment (pazos) [reply plugins/exporter.koplugin/target/json.lua:6]: I think it does depend on the consumer apps. Text processing apps are going to be more prone to catch `text/*` than `application/octet-stream`.

Other apps, like whatsapp, might catch both but treat them differently (the text pasted directly and the binary as an attachment).

And yet another kind of apps, like Dropbox, might deal with any kind of binary and ignore text (the first I know, the second I guess).

File: plugins/exporter.koplugin/target/markdown.lua (hunk 2 above)
[the remark this answers is not in the document]
Comment (pazos) [reply plugins/exporter.koplugin/target/markdown.lua:139]: pinging @uroybd. ^^.

[the remark this answers is not in the document]
Comment (pazos) [reply plugins/exporter.koplugin/target/markdown.lua:139]: whatever. I'll merge this as is. You can make a change in the future if you're still interested :)

File: plugins/exporter.koplugin/base.lua [hunk 3]
```diff
@@ -140,4 +144,14 @@ function BaseExporter:toggleEnabled()
...
 end
 
+--[[--
+Shares text with other apps
+]]
+function BaseExporter:shareText(text, title)
+    local msg_reason = _("Share") .. " " .. self.name
+    local msg_text = type(text) == "string" and text
+    local msg_title = type(title) == "string" and title
+    Device:doShareText(msg_text, msg_reason, msg_title, self.mimetype)
```
Comment (Frenzie) [inline plugins/exporter.koplugin/base.lua:154]: It's no problem is some of these arguments are nil?

**Review by Frenzie — APPROVED** [review]
lgtm

File: plugins/exporter.koplugin/base.lua (hunk 3 above)
Comment (pazos) [reply plugins/exporter.koplugin/base.lua:154]: Nope, the only device that `canShareText` is android.

And android ends calling https://github.com/koreader/android-luajit-launcher/blob/master/app/src/main/java/org/koreader/launcher/MainActivity.kt#L659, where everything other than text is nullable.

**Comment by poire-z** [comment]
What's this PR status ?

**Comment by pazos** [comment]
@poire-z Ready to be merged.

@uroybd can add the title later if he's stil interested.

**Pull request merged**